계산기 [STEP 3] 민쏜

minsson opened the pull request:
안녕하세요, 콘!

Step 3 PR 제출합니다. 검토 부탁드립니다! ☺️
PR이 늦어졌음에도 불구하고 피드백 받을 기회를 주셔서 다시 한번 정말 감사해요...! 😭🙇‍♂️
정말 포기하고 싶지 않았어요...! 🥲 그런 제게 성장통이라고 생각하라고 해주신 말씀이 많은 도움이 되었어요!!

Step 3는 다행히 Step 2에 비해서 좀 더 원활하게 진행이 가능했던 것 같아요.

기능을 UI와 연결하는 게 생각보다 조금 더 어려웠는데, 특히 스크롤 뷰 안에 그동안 입력한 숫자들을 집어 넣는 과정이 처음에는 감이 잡히지 않았었어요. 다행히 잘 해결되었습니다!


# 1. 변수와 상수에 대한 설명
#### `currentOperatorLabel: UILabel!`
숫자입력란의 왼쪽에서 연산자를 표시하는 레이블입니다.

#### `currentNumberLabel: UILabel!`
숫자입력란을 표시하는 레이블입니다.


#### `historyStackView: UIStackView!`
숫자입력내역을 표시해줄 `Vertical Stack View` 입니다. `insertCurrentItemIntoHistoryStackView()` 내부에서 연산자 레이블과 숫자 레이블 2개를 담을 `inputItemStackView`라는 `Horizontal Stack View`를 생성해 여기에 삽입합니다.


#### `resultExpression: String`
`ExpressionParser.parse(from:)`에 넣어 계산 결과를 구하기 위한 `input`으로 사용됩니다. `currentNumber`를 통해 연산자와 숫자를 추가하여 `123+456-789*12`와 같이 하나의 String으로 보관합니다.

#### `firstDecimalPointInCurrentNumber: Bool`
소수점이 한 번의 입력에서 2번 이상 사용되지 않도록 통제하기 위한 변수입니다. 


#### `firstInputAfterCalculation: Bool`
`=` 버튼을 누를 때마다 중복 계산되는 일이 없도록 통제하기 위한 변수입니다. 

#### `numberFormatter: NumberFormatter`
숫자 입력란과 입력내역에서 숫자의 세 자리마다의 쉼표, 최대 자릿수, 반올림을 설정하기 위해 `NumberFormatter`의 인스턴스를 생성해서 사용합니다.



# 2. 프로젝트를 수행하면서 궁금했던 것들

## 1) 전반적인 구조, 설계
코드를 작성하는 전반적인 구조나 설계가 잘 이루어지고 있는지가 정말 많이 궁금해요. 공부를 조금씩 해나가면서 좀 더 넓은 부분을 보면서 개발하고 싶다는 욕심이 나기 시작했는데, 아직은 어디서부터 시작해야할지 감이 잘 오지 않아요. 

저로서는 이런 걸 생각하면 디자인 패턴 정도가 떠오르는데, 아직 기본적인 MVC 패턴부터 더 잘 해야겠다는 생각이 들어 다른 걸 찾아보지는 않았어요. 

혹시 지금은 어떻게 하고 있는지, 그리고 앞으로 어떻게 공부해나가면 좋을지 구체적인 피드백을 주실 수 있으실까요...!?

## 2) 네임스페이스
네임스페이스를 사실상 처음 써봤어요. `enum` 안에 `static let`으로 선언해서 어디에서나 사용할 수 있도록 했는데, 네이밍, 선언 위치 등 이런 식으로 하는 게 맞는지 궁금해요. 


## 3) UI와 로직을 담당하는 변수
`currentNumber`와 `currentNumberLabel`, 그리고 `currentOperator`와 `currentOperatorLabel`가 정의되고 사용되는 방식이 적절한가요?

제가 느끼기에는 사실상 같은 데이터가 들어가는데, 거의 모든 메서드에서 이 두 개의 값을 별도로 설정하고 있다보니 중복된 코드처럼 느껴져요. 혹시 이걸 개선할 수 있는 방법이 있나요? 하나의 변수로 관리할 수는 없을까요?


## 4) 네이밍
전반적으로 네이밍이 직관적으로 잘 읽히는지 궁금해요.
특히, `firstDecimalPointInCurrentNumber: Bool`와
`firstDecimalPointInCurrentNumber: Bool`은 Bool Type이므로 `is, can` 등이 떠올랐지만, 어떤 게 적절한지 선택하기 어려웠어요.


## 5) Scroll View 내부의 Dynamic Stack View
제게는 이 프로젝트에서 최대 난관인 주제였어요.
4-1) 선언한 위치와 사용방법이 적절한지 궁금해요. 같은 기능을 구현하기 위한 더 좋은 설계 방법이 있을지도 궁금해요.
4-2) 지금 앱 기능처럼 스택 뷰에 값이 추가될 때, 어떻게 해야 추가되는 방향을 설정할 수 있는지 궁금해요. 상, 하, 좌, 우, 심지어 중간 등.. 어떻게 하면 통제할 수 있나요? 관련해서 `ios dynamic stack view add position` 등으로 검색해봤지만 찾지 못했어요 😅


## 6) 파일 및 폴더 관리
콘이라면 이 프로그램에서 타입을 어떻게 분류하고, 폴더링 했을지 궁금해요. 저는 아직 명확한 기준이 없어서 그때그때 다르게 분류해서 관리하는 것 같아요. 혹시 참고할만한 레퍼런스가 있을까요?


## 7) 기타
제가 한 질문 외에, 혹시 제가 놓치고 있는 부분이나 알아야할 부분이 있다면, 조언해주시면 감사하겠습니다!! ☺️

minsson committed:
feat: 숫자 입력칸 UI 연동 및 초기값 설정

minsson committed:
feat: 숫자 버튼 탭하면 숫자입력칸에 반영되는 기능 구현

minsson committed:
refactor: CalculatorVC 접근제어 일부 수정 및 가독성 향상을 위한 MARK 주석 기입

minsson committed:
feat: 사칙연산 기호 버튼 탭시 입력칸의 사칙연산 기호가 변경되는 기능 구현

minsson committed:
refactor: 가독성 향상을 위한 변수명 변경

minsson committed:
Merge branch 'ic_6_minsson' into minsson_step3_test

minsson committed:
feat: 현재 입력내용을 화면 상단 입력 내역에 반영해주는 insertCurrentItemIntoHistory() 메서드 구현

minsson committed:
feat: 전체 입력 내용 및 현재 입력 내용을 저장하기 위한 변수 선언, 입력 내역 화면에 입력 반영하는 기능 구현

minsson committed:
fix: UI 상의 사칙연산자 기호와 로직 내 사칙연산자 기호를 동일하게 수정, 사칙연산 버튼 누르고 다른 버튼으로 변경시 반영되지 않던 문제 수정

minsson committed:
feat: 등호 버튼을 누르면 전체 입력내역을 계산하여 화면에 출력하는 기능 구현

이와 관련, 계산 다음에 최초로 operator를 입력 후 숫자를 입력하면 mainLabel 뒤이어 작성되는 문제가 있어, numberButtonTapped 메서드에 firstInputAfterCalculation 추가하여 입력 가능 여부 확인 후 입력 진행

minsson committed:
feat: AC버튼을 누르면 앱의 초기화면으로 돌아가는 기능 구현 (로직 내부 값들도 모두 초기값으로 변경)

minsson committed:
feat: CE 버튼을 누르면 현재 입력 중이던 숫자 또는 연산결과를 삭제하는 기능 구현

minsson committed:
feat: +/- 버튼 누르면 현재 입력한 숫자가 양수, 음수 부호 전환이 되게 하는 기능 구현

minsson committed:
refacotr: 메서드 일부 접근제어 private으로 변경

minsson committed:
feat: 소수점 버튼을 누르면 숫자 입력 레이블과 로직 내 현재 입력 숫자에 반영하는 기능 구현

minsson committed:
feat: UI에 나올 숫자들의 포맷을 지정하기 위한 NumberFormatter의 인스턴스 생성

minsson committed:
refactor: 일부 강제 언래핑을 옵셔널 바인딩을 이용한 언래핑으로 수정

minsson committed:
feat: UI상 모든 숫자 세자리마다 콤마(,)를 자동으로 삽입하는 기능 구현

minsson committed:
refactor: equalButtonTapped(_:) 메서드 내부 구조 가독성 개선

minsson committed:
refactor: 네임스페이스 작성 및 스트링 매직 리터럴 제거

minsson committed:
refactor: 가독성 향상을 위한 불필요한 코드 및 개행 제거

minsson committed:
refactor: 가독성 향상을 위한 개행

minsson committed:
refactor: 에러 문구 매직리터럴 제거

minsson committed:
refactor: 가독성 향상을 위한 변수명, 메서드명 수정

minsson committed:
refactor: 가독성 개선을 위한 변수명 및 개행 수정

minsson committed:
refactor: Formula의 저장속성을 변수에서 상수로 변경

protocorn93 reviewed (commented):
@minsson 고생하셨어요! 너무 조급해하지마세요! 하지만 기한 내 최대한 마무리해보는 것도 실제로 업무를 진행할 때 필요한 부분이라고 생각해요! (절대 나무라는 것 아니니 오해하지마세요! ㅎㅎ🤣)

업무를 진행하는데 있어, 기한을 설정할 것이고 그 안에 어디까지 구현할 수 있겠다! 이런 스케쥴링도 업무를 진행하는데 있어 중요한 기술이니까요~! 

# 궁금했던 점 

> 1) 전반적인 구조, 설계

벌써부터 패턴에 신경쓰지 않으면 좋겠어요. 패턴이란 결국 반복되는 것을 형상화한것이니까요! 아키텍처도 마찬가지에요. 그전에 객체의 역할을 나누는 연습을 해보는게 좋아요! 흔히 말하는 OOP 5원칙을 공부해가면서 이를 코드에 적용시켜보는 것이 그 시작이라고 할 수 있어요. 

그렇게 되면 자연스럽게 의존성 주입이라던지, 의존성 역전의 법칙이라던지 등등 기타 심오한 주제들도 반드시 학습하게 되실거에요! 그렇기 때문에 지금은 역할을 나눠보는, 좀 더 쉽게는 메서드를 더 확실히 나눠보는 연습을 해보시는게 좋을 것 같아요! 

> 2) 네임스페이스

이 부분은 코멘트에 남겨놓았어요! 

> 3) UI와 로직을 담당하는 변수

들어가는 위치와 값이 유사해보여서 그렇게 느껴질 수 있으나 제가 생각했을 때 둘의 역할은 엄연히 다르다고 생각해요. 

오히려 하나의 변수로 관리하면 가독성이 더 떨어질 것 같네요!  🤔

> 4) 네이밍

저는 이와 같은 condition 변수에는 `is-`나 `should-`를 붙이는 편이긴해요! 
(이와는 별개로 최대한 condition 변수는 지양하는게 좋아요! 기억해야할 condition 변수가 많아질 수록 가독성은 떨어지고 유지보수도 힘들어지거든요!  condition 변수는 최대한 메서드의 로컬 변수로 가져가는게 좋아요. 물론 부득이한 경우에는 당연히 사용해야하지만말이죠!)

> 5) Scroll View 내부의 Dynamic Stack View

스크롤뷰가 확실히 다루기 까다롭죠! 먼저 frame layout과 content layout에 대해 공부해보시는 것도 좋아요! 그리고 https://fluffy.es/scrollview-storyboard-xcode-11/ 이 블로그가 인터페이스 빌더로 스크롤뷰 설정하는 법에 대해 친절히 나름 잘 설명하고 있어서 한번 읽어보시는 것도 추천해드려요! 

> 6) 파일 및 폴더 관리

이 영역도 사실 굉장히 주관적인 영역이에요! 이 부분은 다른 유명한 사람들의 github 프로젝트 구조를 보시고 참고해보시는 것도 좋아요! 저는 전수열님의 프로젝트 구조를 차용해서 사용하곤해요! 

https://github.com/devxoul/RxTodo 이 레포를 참고해봐도 좋구요!

minsson committed:
refactor: 타입의 네이밍 수정

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+    static let zero = "0"
+}
+
+final class CalculatorVC: UIViewController {
```
어디서 보고 짧아서 쓰기도, 보기도 편하겠다 싶어서 줄여봤는데 안 줄이는 게 좋겠군요..!! 😅

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+    private var firstDecimalPointInCurrentNumber = true
+    private var firstInputAfterCalculation = false
+    
+    let numberFormatter: NumberFormatter = {
```
접근 제어자를 빠뜨렸네요..! 수정했습니다!

minsson committed:
refactor: 일부 변수 및 메서드 접근제어 수준 변경

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+
+import UIKit
+
+enum NameSpace {
+    static let maximumDigits = 20
+    static let minus: Character = "-"
+    
+    static let emptyString = ""
+    static let comma = ","
+    static let dot = "."
+    static let zero = "0"
+}
```
네이밍이 모호하다고 생각했는데, 그 이유가 변수들의 역할이 달라서였네요!
그리고 다른 곳에서 안 쓰니 `CalculatorViewController` 안에 `private enum`으로 넣는 것도 너무 좋은 것 같아요!

콘의 피드백을 듣고 처음에는 `maximumDigits`와 나머지 2개로 나눠보려고 했는데, 
더 생각해보니 아래와 같이 3가지로 나누는 건 어떨까 하는 생각이 들었어요.

(기존에 있던 dot은 버튼의 currentTitle을 받아오는 게 나을 것 같아 제거했습니다!)

```swift
    private enum LogicConstraintConstants {
        static let maximumDigits = 20
    }
    
    private enum LogicConstants {
        static let emptyString = ""
        static let comma = ","
        static let minus: Character = "-"
    }
    
    private enum UIConstants {
        static let emptyString = ""
        static let zero = "0"
    }
```

이 때, 몇 가지 고민이 있었어요.

1) 네임 스페이스를 하나의 클래스 안에서 이렇게 여러 개로 두는 게 일반적인 것인지
2) 위의 코드에서 분류상 `emptyString`, 즉, `공백`은 UI와 로직 모두에서 사용되는데, 똑같은 `공백`을 두 개의 네임 스페이스에 나눠서 놓는 게 맞는 건지, 그리고 이런 게 고민인 상황이라면, 애초에 Logic과 UI로 나누는 분류 방법 자체가 좋지 않은 방향인 건지
3) `LogicConstraintConstants`에 있는 `maximumDigits = 20`는 네임스페이스로 분류해놓는 거 자체가 적절한 것인지, 혹시 한번만 쓰이는 것이므로 네임스페이스를 쓰지 않고 사용되는 곳에서 그냥 매직넘버로 쓰는 게 나은지

콘은 어떻게 생각하시나요?

protocorn93 replied:
1. 저는 클래스 내부에 여러 네임스페이스를 정의하기도 해요! Typo, Metric 같이!
2. 음.. 이건 저도 고민이긴하네요. 다만 현재 상태에선 각각 가져가는 것도 괜찮아보여요. 추후 유지보수가 어렵다 느껴질 때 공통 레벨로 빼는게 좋아보이네요. 
3. 저는 그런 값들도 네임스페이스로 정의하기는 해요

minsson replied:
콘 덕분에 네임스페이스와 실전적인 활용법에 대해 더 잘 이해할 수 있었어요! 감사합니다!!

minsson committed:
refactor: 네임스페이스 구조, 네이밍 변경 및 코드에 적용

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        currentOperatorLabel.text = NameSpace.emptyString
+        currentNumberLabel.text = NameSpace.zero
```
아래와 같이 메서드로 분리해봤어요!
말씀해주신 것에 더해 로직 관련 부분에도 비슷한 게 있어서 마찬가지로 분리해봤어요.

그런데 첫번째 메서드에서 `MainLabel` 이라는 표현이 적절한지 고민이에요. 
더 구체적이고 명확하게 `resetCurrentOperatorAndNumberLabel` 이라는 표현이 더 나았을까요?

```swift
    private func resetCurrentMainLabel() {
        currentOperatorLabel.text = UIConstants.emptyString
        currentNumberLabel.text = UIConstants.zero
    }
    
    private func resetCurrentOperatorAndNumber() {
        currentOperator = LogicConstants.emptyString
        currentNumber = LogicConstants.emptyString
    }
```

minsson replied:
늘 네이밍을 가급적 구체적으로 하려고 했었는데, 콘이 제안해주신 `resetLabels`라는 네이밍에 대해 생각해보니, 어차피 하나의 뷰컨트롤러 안에서만 쓰일 함수이므로 `Labels`라는 단어 이상으로 구체화하지 않아도 될 것 같다는 생각이 들었어요! `resetLabels`로 수정했습니다!

minsson committed:
refactor: 현재 메인 레이블 및 로직 내 값 초기화하는 코드를 메서드로 분리

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+
+enum NameSpace {
+    static let maximumDigits = 20
+    static let minus: Character = "-"
```
아래의 `switchPositiveNegativeButtonTapped(_:)` 메서드에서 `currentNumber.first == LogicConstants.minus` 라는 코드가 있어요. 현재 로직 내의 숫자 값의 첫 글자가 `-`인지, 즉, 음수로 전환된 상태인지를 확인하는 로직이에요.

이 때 사용되는 좌항의 `currentNumber.first`에서 `.first`의 리턴은 `Character`이므로, 타입을 맞춰줬어요.

아래는 해당 메서드의 코드 전문이에요.
```swift
    @IBAction private func switchPositiveNegativeButtonTapped(_ sender: UIButton) {
        if currentNumber == LogicConstants.emptyString {
            return
        } else if currentNumber.first == LogicConstants.minus {
            currentNumber.remove(at: currentNumber.startIndex)
        } else {
            currentNumber = String(LogicConstants.minus) + currentNumber
        }
     
        currentNumberLabel.text = currentNumber
    }
```


`static let minus: Character = "-"`를 `static let minus: String = "-"`으로 바꾸면 아래와 같이 오류가 발생해요.

<img width="916" alt="스크린샷 2022-06-10 14 20 58" src="https://user-images.githubusercontent.com/96630194/172995837-29adddd1-4e90-4d64-9627-81cd151bfcbe.png">

minsson committed:
refactor: 숫자 입력 내역 스택뷰에 아이템을 추가하는 메서드 구조 변경

- 객체가 currentNumber, currentOperator 변수를 갖고있는것과 별개로 이 메서드는 number와 operator를 받도록 수정
- 객체 상태에 접근하는 곳을 줄여 휴먼 에러 발생 가능성과 디버깅 난이도를 낮추기 위함
- 이 메서드의 경우 필요한 순간 그 당시의 값을 이용해 뷰를 만드는 것이므로, 아규먼트로 값을 받아 처리하면 역할을 분명히 하고, 잘못 사용될 가능성도 낮출 수 있음

minsson committed:
chore: 아규먼트 리스트 내 불필요한 escape 제거

minsson committed:
refactor: condition 변수 네이밍 변경

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+            currentNumberLabel.text = CalculatorError.emptyCalculatorItemQueue.description
+        }
+        
+        insertCurrentItemIntoHistoryStackView()
+        updateCurrentNumberLabelWithFormat()
+        
+        currentOperatorLabel.text = NameSpace.emptyString
+        
+        currentOperator = NameSpace.emptyString
+        currentNumber = NameSpace.emptyString
+        
+        firstInputAfterCalculation = true
+        firstDecimalPointInCurrentNumber = true
```
구체적인 예시, 심지어 제가 쓴 코드를 통해 설명해주시니 `추상화`, `추상화 레벨`이라는 개념을 조금 더 잘 이해하게 된 것 같아요! 감사합니다, 콘! 😁

코드는 다른 코멘트 내용까지 반영해 우선 아래와 같이 고쳐봤어요.

```swift
        insertItemIntoHistoryStackView(operator: currentOperator, number: currentNumber)
        updateCurrentNumberLabelWithFormat()
        
        currentOperatorLabel.text = UIConstants.emptyString
        
        resetCurrentOperatorAndNumber()
        isFirstInputAfterCalculation = true
        isFirstDecimalPointInCurrentNumber = true
```

그런데 여기서 질문이 있어요!
`currentOperatorLabel.text = UIConstants.emptyString`
`isFirstInputAfterCalculation = true
`isFirstDecimalPointInCurrentNumber = true`

이 세개는 메서드로 묶기가

minsson replied:
콘의 조언대로 추상화 레벨을 맞춰보고자 한 줄짜리 코드 또한 메서드로 분리해봤어요. 확실히 추상화 레벨을 맞추고, 어떤 기능을 하는 메서드인지 네이밍을 통해 드러내니 훨씬 더 가독성이 좋아졌다는 생각이 들었어요.

아래 코드는 콘이 이 코멘트에서 짚어주신 부분이에요! 콘이 처음 짚어주신 부분과, 조언을 따른 후의 전후 코드를 비교해보면 확실히 차이가 느껴지네요..!!!

```swift
        insertItemIntoHistoryStackView(operator: currentOperator, number: currentNumber)
        updateCurrentNumberLabelWithFormat()
        
        eraseCurrentOperatorLabel()
        
        resetCurrentOperator()
        resetCurrentNumber()
        
        allowNumberInput()
        allowDecimalPointInput()
```

minsson committed:
refactor: 가독성 향상을 위한 개행 수정

minsson committed:
refactor: 에러를 캐스팅하여 description에 바로 접근하도록 수정

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+        UIStackView.animate(withDuration: 0.3) {
+            inputItemStackView.isHidden = false
+        }
+    }
```
지금 해주신 피드백은 앞으로 구조를 짜기 전 늘 미리 고민해야 할 부분인 것 같아요.

말씀해주신 것처럼 애초에 이런 변수가 없으면 좋겠지만, 지금처럼 결론적으로는 그 변수가 있을 때도 `객체가 currentNumber, currentOperator 변수를 갖고있는것과 별개로 이 메서드는 number와 operator를 받는건 어떨까요?` 라고 해주신 부분도 신기했어요.

그런 파라미터가 없어도 코드가 정상적으로 작동하지만, 메서드의 역할도 더 분명히 하고, 휴먼 에러를 낮추기 위해 만들어준다는 게 무척 세심한 예방법이라는 생각이 들었어요..!

아래와 같이 수정해봤어요! 콘의 의도를 잘 이해한 걸까요?

```swift
    private func insertItemIntoHistoryStackView(`operator`: String, number: String) {
        let inputItemStackView = UIStackView()
        inputItemStackView.translatesAutoresizingMaskIntoConstraints = false
        inputItemStackView.axis = .horizontal
        inputItemStackView.alignment = .fill
        inputItemStackView.distribution = .fill
        inputItemStackView.spacing = 8
        
        let operatorSignLabel = UILabel()
        operatorSignLabel.translatesAutoresizingMaskIntoConstraints = false
        operatorSignLabel.textColor = .white
        operatorSignLabel.backgroundColor = .black
        operatorSignLabel.text = `operator`
        operatorSignLabel.font = UIFont.preferredFont(forTextStyle: .title3)
        
        let numberLabel = UILabel()
        numberLabel.translatesAutoresizingMaskIntoConstraints = false
        numberLabel.textColor = .white
        numberLabel.backgroundColor = .black
        numberLabel.font = UIFont.preferredFont(forTextStyle: .title3)
        let numberForFormatting = Double(number.replacingOccurrences(of: LogicConstants.comma, with: LogicConstants.emptyString))
        numberLabel.text = numberFormatter.string(for: numberForFormatting)
        
        inputItemStackView.addArrangedSubview(operatorSignLabel)
        inputItemStackView.addArrangedSubview(numberLabel)
        
        inputItemStackView.isHidden = true
        historyStackView.addArrangedSubview(inputItemStackView)
        
        UIStackView.animate(withDuration: 0.3) {
            inputItemStackView.isHidden = false
        }
```

minsson commented on Calculator/Calculator/Controller/CalculatorVC.swift:
```diff
@@ -0,0 +1,206 @@
+            let result = try ExpressionParser.parse(from: resultExpression).result()
+            currentNumberLabel.text = String(result)
+        } catch CalculatorError.dividedByZero {
+            currentNumberLabel.text = CalculatorError.dividedByZero.description
```
아래와 같이 고쳐봤는데, 콘이 의도하신 바가 맞을까요?
(그리고 타입캐스팅을 이번에 처음 써봐서, 이런 식으로 활용할 수 있다는 게 무척 신기했어요!)

```swift
        do {
            let result = try ExpressionParser.parse(from: resultExpression).result()
            currentNumberLabel.text = String(result)
        } catch let error as CalculatorError {
            currentNumberLabel.text = error.description
        } catch {
        }
```

다만 질문이 있는데, `catch let error as CalculatorError`에서 모든 케이스를 처리해주니, 밑에 `catch` 블럭은 쓰지 않아도 된다고 생각해 삭제해봤어요. 그랬더니 모든 케이스가 다뤄지지 않고 있다는 에러가 났어요.

그래서 일단은 `catch` 블럭은 넣어놓고, 안에는 내용 없이 비워놨는데, 이 모습이 부자연스럽다는 생각이 들어요. 혹시 더 좋은 방법이 있을까요?

protocorn93 replied:
네네! 보통 그 곳에는 unexpectedError 에러 케이스로 처리하긴해요.

minsson replied:
아래와 같이 바꿔봤는데 적절한지 확신이 없습니다..!

```swift
do {
            let result = try ExpressionParser.parse(from: resultExpression).result()
            currentNumberLabel.text = String(result)
        } catch let error as CalculatorError {
            currentNumberLabel.text = error.description
        } catch {
            currentNumberLabel.text = "Unexpected error: \(error)"
        }
```

minsson committed:
style: 프로젝트 내 파일 및 폴더의 구조 변경

minsson committed:
Fix: info.plist 경로 오류를 해결하기 위한 파일 위치 변경

minsson committed:
Merge branch 'yagom-academy:ic_6_minsson' into ic_6_minsson

minsson committed:
refactor: 가독성 개선을 위한 메서드명 변경

minsson committed:
refactor: 가독성 개선을 위한 파일 내 메서드 분류 및 위치 변경

minsson committed:
refactor: 가독성 개선을 위한 메서드의 추상화 레벨 통일

minsson commented:
@protocorn93 
안녕하세요, 콘!

피드백해주신 내용은 모두 수정했고, 개별코멘트에 답변도 남겨 놓았어요.
추가 수정사항이 없다면 머지 희망합니다!

계산기 프로젝트는 제게 정말 많은 교훈을 준 프로젝트였어요. 
코딩 실력도, 그리고 정신적으로도 성장할 수 있는 계기가 되었던 것 같아요.

또, 마음고생을 너무 많이 해서 영원히 못 잊을 것 같기도 하고요 🥲
이 모든 과정에 콘이 함께해주셔서 정말 감사했습니다...!!! 🙇‍♂️
